packages/console/src/hooks/useFileDecryption.ts

```diff
@@ -0,0 +1,153 @@
+import { useW3 } from '@storacha/ui-react'
+import { useState } from 'react'
+import type { Space, UnknownLink } from '@storacha/ui-react'
+import { parse as parseLink } from 'multiformats/link'
+import { create as createEncryptedClient } from '@storacha/encrypt-upload-client'
+import { useKMSConfig } from '@storacha/ui-react'
+import { decrypt } from '@storacha/capabilities/space'
+import type { FileMetadata } from '@storacha/encrypt-upload-client/types'
+
+interface DecryptionState {
+  loading: boolean
+  error: string | null
+  fileMetadata?: FileMetadata
+}
+
+export const useFileDecryption = (space?: Space) => {
+  const [{ client }] = useW3()
+  const [state, setState] = useState<DecryptionState>({
+    loading: false,
+    error: null,
+    fileMetadata: undefined
+  })
+
+  const { createKMSAdapter, isConfigured } = useKMSConfig()
+
+  const downloadBlob = (blob: Blob, filename: string) => {
+    const url = URL.createObjectURL(blob)
+    const a = document.createElement('a')
+    a.href = url
+    a.download = filename
+    document.body.appendChild(a)
+    a.click()
+    document.body.removeChild(a)
+    URL.revokeObjectURL(url)
+  }
+
+  const streamToBlob = async (stream: ReadableStream, mimeType?: string): Promise<Blob> => {
+    const reader = stream.getReader();
+    const chunks: Uint8Array[] = [];
+    let done = false;
+    let totalBytes = 0;
+
+    while (!done) {
+      const { value, done: isDone } = await reader.read();
+      done = isDone;
+      if (value) {
+        chunks.push(value);
+        totalBytes += value.length;
+      }
+    }
+
+    return new Blob(chunks, { type: mimeType || 'application/octet-stream' });
+  };
+
+  const decryptAndDownload = async (cid: UnknownLink | string, filename?: string) => {
+    if (!client || !space || space.access?.type !== 'private') {
+      throw new Error('Invalid state: client, space, or private space access required')
+    }
+
+    setState({ loading: true, error: null, fileMetadata: undefined })
+
+    try {
+      // Create crypto adapter using shared KMS config
+      const cryptoAdapter = await createKMSAdapter()
```

Comment:
nb: is there any startup cost to recreating the encrypted client each time? Would it make sense to have a seperate "useEncryptedClient" or something like that?

Reply:
Yeah, there is, but it is minimal (~few milliseconds of object creation and URL validation). You don't even notice it on the UI. We can certainly have a new hook for that. But I'll leave optimizations for a second iteration if you don't mind.